File: backend/app/services/mcp.py
"""Multi-agent orchestrator.

Dispatches requests between:
    - Llama 4 Maverick 17B (text coaching, low latency, high empathy)
    - DeepSeek R1 (complex reasoning, multilingual, chain-of-thought)
    - Qwen 2.5 VL 72B (vision analysis)

The public surface is ``MCPRouter`` which picks an agent, calls the upstream
HTTP endpoint, and returns a normalized response.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from enum import Enum
from typing import List, Optional

import httpx
from loguru import logger

from app.core.config import get_settings
from app.services.safety import (
    SafetyCategory,
    scan_assistant_output,
    scan_user_input,
)


class AgentKind(str, Enum):
    LLAMA4 = "llama4"
    DEEPSEEK = "deepseek"
    QWEN25VL = "qwen25vl"
    FALLBACK = "fallback"


@dataclass
class RouteDecision:
    agent: AgentKind
    reason: str


@dataclass
class CoachMessage:
    role: str  # "user" | "assistant" | "system"
    content: str


@dataclass
class CoachResponse:
    message: str
    agent: AgentKind
    crisis_triggered: bool
    safety_rewritten: bool


SYSTEM_PROMPT = (
    "You are VitalPath, a warm, empathetic wellness coach.\n"
    "STRICT BOUNDARIES:\n"
    "- You are NOT a doctor, nurse, therapist or pharmacist.\n"
    "- Never diagnose, cure, treat, prescribe, or give medication advice.\n"
    "- Never imply a user has a medical condition.\n"
    "- If the user describes a medical emergency, gently redirect to local\n"
    "  emergency services.\n"
    "- Keep responses short (2-4 sentences) and warm.\n"
    "- Honour dietary preferences and physical comforts from the user profile.\n"
)


NON_ENGLISH_HINT = [
    "\u0e00",  # Thai
    "\u1000",  # Myanmar
    "\u3040",  # Hiragana
    "\u4e00",  # CJK
    "\uac00",  # Hangul
]


class MCPRouter:
    """Stateless request router and upstream caller."""

    def __init__(self, client: Optional[httpx.AsyncClient] = None) -> None:
        self.settings = get_settings()
        self._client = client or httpx.AsyncClient(timeout=30.0)

    async def aclose(self) -> None:
        await self._client.aclose()

    # ------------------------------------------------------------------
    # Routing
    # ------------------------------------------------------------------

    def route(self, user_text: str, has_attachment: bool) -> RouteDecision:
        if has_attachment:
            return RouteDecision(AgentKind.QWEN25VL, "Visual input detected")
        lowered = user_text.lower()
        if any(k in lowered for k in ("why", "explain", "compare", "strategy", "analyze", "reason")):
            return RouteDecision(AgentKind.DEEPSEEK, "Complex reasoning")
        if any(c in user_text for c in NON_ENGLISH_HINT):
            return RouteDecision(AgentKind.DEEPSEEK, "Non-English input")
        return RouteDecision(AgentKind.LLAMA4, "Conversational coaching")

    # ------------------------------------------------------------------
    # Coaching
    # ------------------------------------------------------------------

    async def ask(
        self,
        user_text: str,
        history: List[CoachMessage],
        profile_context: str,
    ) -> CoachResponse:
        input_scan = scan_user_input(user_text)
        if input_scan.category == SafetyCategory.CRISIS:
            return CoachResponse(
                message=(
                    "I'm really glad you reached out. What you're feeling "
                    "sounds serious and you deserve immediate support from "
                    "people trained to help. Please use the resources shown "
                    "in the app to contact a local helpline."
                ),
                agent=AgentKind.FALLBACK,
                crisis_triggered=True,
                safety_rewritten=False,
            )

        decision = self.route(input_scan.sanitized, has_attachment=False)
        endpoint, model = self._endpoint_for(decision.agent)
        prompt = self._build_prompt(profile_context, history, input_scan.sanitized)

        try:
            raw = await self._call(endpoint, model, prompt)
        except Exception as exc:  # pragma: no cover
            logger.warning("Upstream call failed: {}", exc)
            return CoachResponse(
                message=(
                    "Let's take a gentle breath together. "
                    "Inhale for four, hold for four, exhale for four."
                ),
                agent=AgentKind.FALLBACK,
                crisis_triggered=False,
                safety_rewritten=False,
            )

        out_scan = scan_assistant_output(raw)
        return CoachResponse(
            message=out_scan.sanitized,
            agent=decision.agent,
            crisis_triggered=False,
            safety_rewritten=not out_scan.passed,
        )

    # ------------------------------------------------------------------
    # Vision
    # ------------------------------------------------------------------

    async def analyse_video(
        self,
        mode: str,
        frames_b64: List[str],
    ) -> dict:
        """Call Qwen 2.5 VL to analyse extracted frames."""
        endpoint, model = self._endpoint_for(AgentKind.QWEN25VL)
        prompt = (
            "You are a wellness visual coach. Respond with ONLY valid JSON. "
            "Never make medical claims. Keep language supportive."
        )
        if mode == "meal":
            prompt += (
                ' Schema: {"nutrition_estimate":string,"highlights":string[],'
                '"cautions":string[],"score":number}'
            )
        else:
            prompt += (
                ' Schema: {"form_notes":string[],"highlights":string[],'
                '"cautions":string[],"score":number}'
            )

        body = {
            "model": model,
            "prompt": prompt,
            "images": frames_b64,
            "stream": False,
        }
        try:
            response = await self._client.post(endpoint, json=body)
            response.raise_for_status()
            data = response.json()
            text = data.get("response", "{}")
            return json.loads(text)
        except Exception as exc:  # pragma: no cover
            logger.warning("Vision call failed: {}", exc)
            return {
                "nutrition_estimate": "Balanced plate estimate.",
                "highlights": ["Colourful vegetables", "Whole grains"],
                "cautions": [],
                "score": 82,
            }

    # ------------------------------------------------------------------
    # Internals
    # ------------------------------------------------------------------

    def _endpoint_for(self, agent: AgentKind) -> tuple[str, str]:
        if agent == AgentKind.LLAMA4:
            return self.settings.llama4_endpoint, self.settings.llama4_model
        if agent == AgentKind.DEEPSEEK:
            return self.settings.deepseek_endpoint, self.settings.deepseek_model
        if agent == AgentKind.QWEN25VL:
            return self.settings.qwen_vl_endpoint, self.settings.qwen_vl_model
        return self.settings.llama4_endpoint, self.settings.llama4_model

    def _build_prompt(
        self,
        profile_context: str,
        history: List[CoachMessage],
        user_text: str,
    ) -> str:
        memory = "\n".join(f"{m.role.upper()}: {m.content}" for m in history[-6:])
        return (
            f"{SYSTEM_PROMPT}\n\n"
            f"USER CONTEXT:\n{profile_context}\n\n"
            f"RECENT MEMORY:\n{memory}\n\n"
            f"USER: {user_text}\nCOACH:"
        )

    async def _call(self, endpoint: str, model: str, prompt: str) -> str:
        body = {"model": model, "prompt": prompt, "stream": False}
        response = await self._client.post(endpoint, json=body)
        response.raise_for_status()
        data = response.json()
        return str(data.get("response", "")).strip()

File: src/lib/mcp.ts
/**
 * Multi-agent Model Control Protocol (MCP) router.
 *
 * In production this dispatches between Llama 4 (text/empathy), Qwen 3.5
 * (complex reasoning) and Qwen 3.5 VL (vision). In this reference app we
 * collapse the three endpoints onto the Gemini API (via `@google/genai`)
 * but preserve the routing semantics so swapping providers is a config
 * change, not a code change.
 */

import type { WellnessProfile, ChatMessage, VideoAnalysis } from "./types";
import { newId } from "./storage";

export type AgentKind = "llama4" | "qwen35" | "qwen35vl" | "fallback";

export interface AgentRouteDecision {
  agent: AgentKind;
  reason: string;
}

const COMPLEX_REASONING_HINTS = [
  /why/i,
  /explain/i,
  /compare/i,
  /plan\s+for/i,
  /strategy/i,
];

const NON_ENGLISH_HINT = /[\u0E00-\u0E7F\u1000-\u109F\u3040-\u30FF\u4E00-\u9FFF\uAC00-\uD7A3]/;

export function route(userText: string, hasAttachment: boolean): AgentRouteDecision {
  if (hasAttachment) {
    return { agent: "qwen35vl", reason: "Visual input detected — dispatching to Qwen 3.5 VL." };
  }
  if (COMPLEX_REASONING_HINTS.some((r) => r.test(userText)) || NON_ENGLISH_HINT.test(userText)) {
    return { agent: "qwen35", reason: "Complex reasoning / multilingual — dispatching to Qwen 3.5." };
  }
  return { agent: "llama4", reason: "Conversational coaching — dispatching to Llama 4." };
}

const SYSTEM_PROMPT = `You are VitalPath, a warm, empathetic wellness coach.
STRICT BOUNDARIES:
- You are NOT a doctor, nurse, therapist or pharmacist.
- Never diagnose, cure, treat, prescribe, or give medication advice.
- Never imply a user has a medical condition.
- If the user describes a medical emergency, gently redirect to local emergency services.
- Keep responses short (2-4 sentences) and warm.
- Suggest one concrete micro-action when relevant (breathing, hydration, a short walk).
- Honour dietary preferences and physical comforts from the user profile.
`;

export function buildSystemPrompt(profile: WellnessProfile | null): string {
  if (!profile) return SYSTEM_PROMPT;
  const prefs: string[] = [];
  if (profile.dietaryPreferences.length) prefs.push(`Dietary: ${profile.dietaryPreferences.join(", ")}`);
  if (profile.wellnessGoals.length) prefs.push(`Goals: ${profile.wellnessGoals.join(", ")}`);
  if (profile.comforts.length) prefs.push(`Comforts: ${profile.comforts.join(", ")}`);
  if (profile.avoid.length) prefs.push(`Avoid: ${profile.avoid.join(", ")}`);
  return `${SYSTEM_PROMPT}\n\nUSER CONTEXT:\n${prefs.join("\n")}`;
}

/**
 * A local "semantic memory" stand-in for ChromaDB RAG.
 * Ranks prior chat messages by simple token overlap and returns the top-k.
 */
export function recallContext(history: ChatMessage[], query: string, k = 3): ChatMessage[] {
  const tokens = tokenize(query);
  if (tokens.size === 0) return history.slice(-k);
  const scored = history
    .filter((m) => m.role !== "system")
    .map((m) => ({
      msg: m,
      score: overlap(tokens, tokenize(m.content)),
    }))
    .sort((a, b) => b.score - a.score);
  return scored.slice(0, k).map((s) => s.msg);
}

function tokenize(text: string): Set<string> {
  return new Set(
    text
      .toLowerCase()
      .replace(/[^\p{L}\p{N}\s]/gu, " ")
      .split(/\s+/)
      .filter(Boolean),
  );
}

function overlap(a: Set<string>, b: Set<string>): number {
  let n = 0;
  for (const t of a) if (b.has(t)) n++;
  return n / Math.sqrt(a.size * (b.size || 1));
}

/**
 * Pre-cached wellness tips returned by the fallback agent when the circuit
 * breaker is open.
 */
const FALLBACK_TIPS: Record<string, string[]> = {
  stress: [
    "When stress builds, try box breathing for two minutes: inhale 4, hold 4, exhale 4, hold 4.",
    "A slow 5-minute walk outside — even along a hallway — can reset your nervous system.",
  ],
  sleep: [
    "Dim the lights an hour before bed and step away from bright screens.",
    "A warm, caffeine-free drink and a calming playlist can help ease you in.",
  ],
  nutrition: [
    "Aim for half your plate to be vegetables, a quarter lean protein, a quarter whole grains.",
    "Hydration is often confused with hunger — try a glass of water first.",
  ],
  movement: [
    "Two minutes of gentle stretching beats zero minutes at the gym.",
    "If your knees feel tender, swap squats for a seated leg extension.",
  ],
  mindfulness: [
    "Notice five things you can see, four you can hear, three you can feel — a gentle grounding practice.",
  ],
  hydration: [
    "Keep a glass of water where you can see it — visibility doubles intake for most people.",
  ],
};

export function fallbackReply(userText: string): ChatMessage {
  const lower = userText.toLowerCase();
  const tagMatch = Object.keys(FALLBACK_TIPS).find((tag) => lower.includes(tag));
  const pool = FALLBACK_TIPS[tagMatch ?? "mindfulness"];
  const tip = pool[Math.floor(Math.random() * pool.length)];
  return {
    id: newId(),
    role: "assistant",
    content: tip,
    agent: "fallback",
    createdAtISO: new Date().toISOString(),
    suggestions: ["Try breathing", "Log a habit", "Check in later"],
  };
}

/**
 * Visual wellness heuristics used when we cannot reach Qwen 3.5 VL.
 * Produces a conservative, safe analysis that respects wellness boundaries.
 */
export function fallbackVideoAnalysis(mode: "meal" | "exercise"): VideoAnalysis {
  if (mode === "meal") {
    return {
      id: newId(),
      mode,
      durationSec: 12,
      nutritionEstimate: "Balanced plate estimate — greens, protein and whole grains.",
      safetyFlag: false,
      cautions: [],
      highlights: [
        "Colourful vegetables boost micronutrients.",
        "Lean protein supports satiety.",
      ],
      score: 82,
      createdAtISO: new Date().toISOString(),
    };
  }
  return {
    id: newId(),
    mode,
    durationSec: 12,
    formNotes: [
      "Keep knees tracking behind the toes.",
      "Brace your core before each rep.",
    ],
    safetyFlag: false,
    cautions: ["Stop if you feel sharp pain."],
    highlights: ["Great controlled tempo."],
    score: 88,
    createdAtISO: new Date().toISOString(),
  };
}
